Add design principles vignette (#38)
epiverse-trace/epichains: vignettes/design-principles.Rmd

---
title: "Design Principles for {epichains}"
subtitle: "An R package for simulating, handling, and analysing transmission chains in R"
pkgdown:
  as_is: true
bibliography: references.json
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Design Principles for {epichains}}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette outlines the design decisions that have been taken during the
development of the `{epichains}` R package, and provides some of the reasoning,
and possible pros and cons of each decision.

The goal here is to make it easy to acquaint oneself with the code base
in the absence of the current maintainer. This will ease future contributions
and maintenance of the package.

## Scope

{epichains} aims to facilitate:

* Simulation of transmission chains using branching processes methods,

> * Handling of transmission chains using objected-orientated design, and

Reviewer: ```suggestion
* Handling of transmission chains with an easily accessible user interface, and
```

I think OOP is more lower level

by varying the inputs. Often, in cases where there is need for more than one
run of the model and/or with more than one set of parameter values, the inputs
and outputs are stored in separate data structures. However, this approach can
be limiting when performing scenario analyses, as the user has to manually
manipulate the two objects with reshaping and joining operations. This has
the potential to lead to errors and loss of information. Hence,
`simulate_chains()` and `simulate_summary()` return the dedicated classes
`<epichains>` and `<epichains_summary>` objects respectively that store the
input parameters and the output of the simulation in a single object.

The `<epichains>` class extends the `<data.frame>`, using columns
to store information about the simulated transmission chains and the
parameter values as attributes. `<data.frame>` was chosen because its tabular
structure allows us to store information in rows and columns, and is a widely
used data structure in R. Similarly, the `<epichains_summary>` class is a
superclass (an extension) of R's `<numeric>` class and stores the parameter
values as attributes.

> The `<epichains>` object contains information about the whole outbreak, but

Reviewer: ```suggestion
The `<epichains>` object contains information about the whole simulated outbreak (potentially with multiple simulations), but
```

key summaries are not easily deduced from a quick glance of the object. Hence,
the class has a dedicated `format()/print()` method to print the simulated
transmission chains in a manner similar to a `<dataframe>`, but accompanied
by extra summary information including the number of chains simulated, number
of generations reached, and the number of unique infectors created. These
summaries are useful for quickly assessing the output of the simulation.

> Importantly, the `<epichains>` class has a `summary()` method that returns

Reviewer: ```suggestion
The `<epichains>` class has a `summary()` method that returns
```

an `<epichains_summary>` object. This is a design decision that was taken to
allow for easy coercion between an infection tree obtained from
`simulate_chains()` and summaries of the infection tree otherwise attainable
by a separate run of `simulate_summary()` with the same parameter values passed
to `simulate_chains()`.

Lastly, `<epichains>` objects have an `aggregate()` method that aggregates the
transmission tree into cases by "generation" or "time". This is syntactic sugar
for the `dplyr::group_by()` and `dplyr::summarise()` style of aggregation.

In summary, an `<epichains>` object has the following structure:

* Columns:
  * `infectee_id` (`<numeric>`)
  * `infector_id` (`<numeric>`)

>   * `generation` (`<numeric>`), and optionally,

Reviewer: are these numeric or integer? (really they should be integer)

Author: They should be. They start off as integer but thanks to R's type instability, they end up being numeric. I think there's an issue to resolve this here: https://github.com/epiverse-trace/epichains/issues/136.

>   * `infectee_id` (`<numeric>`)
>   * `infector_id` (`<numeric>`)
>   * `generation` (`<numeric>`), and optionally,

Author: ```suggestion
  * `infectee_id` (`<integer>`)
  * `infector_id` (`<integer>`)
  * `generation` (`<integer>`), and optionally,
```
